Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift

```diff
@@ -0,0 +1,414 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2023 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import SymbolKit
+
+// MARK: From symbols
+
+extension PathHierarchy {
+    /// Returns the lists of the type names for the symbol's parameters and return values.
+    static func functionSignatureTypeNames(for symbol: SymbolGraph.Symbol) -> (parameterTypeNames: [String], returnTypeNames: [String])? {
+        guard let signature = symbol[mixin: SymbolGraph.Symbol.FunctionSignature.self] else {
+            return nil
+        }
+        
+        return (
+            signature.parameters.compactMap { parameterTypeSpellings(for: $0.declarationFragments) },
+            returnTypeSpellings(for: signature.returns).map { [$0] } ?? []
+        )
+    }
+    
+    private static func parameterTypeSpellings(for fragments: [SymbolGraph.Symbol.DeclarationFragments.Fragment]) -> String? {
+        typeSpellings(for: fragments)
+    }
+    
+    private static func returnTypeSpellings(for fragments: [SymbolGraph.Symbol.DeclarationFragments.Fragment]) -> String? {
+        if fragments.count == 1, knownVoidReturnValues.contains(fragments.first!) {
+            // We don't want to list "void" return values as type disambiguation
+            return nil
+        }
+        return typeSpellings(for: fragments)
+    }
+    
+    private static let knownVoidReturnValues = ParametersAndReturnValidator.knownVoidReturnValuesByLanguage.flatMap { $0.value }
+    
+    private static func typeSpellings(for fragments: [SymbolGraph.Symbol.DeclarationFragments.Fragment]) -> String? {
+        var accumulated = ""[...]
+        
+        for fragment in fragments {
+            switch fragment.kind {
+            case .identifier where fragment.preciseIdentifier == nil, // Skip the argument label
+                 .keyword,   // Skip keywords ("inout", "consuming", "each", etc.)
+                 .attribute, // Skip attributes ("@escaping", custom result builders, etc.)
+                 .numberLiteral, .stringLiteral, // Skip literals
+                 .externalParameter, .genericParameter, .internalParameter:
+                continue
+
+            default:
```

Comment:
Should we be more explicit and only capture the fragment kinds we care about, that would make me feel a bit better if we start seeing unexpected values.

Reply:
Updated in https://github.com/swiftlang/swift-docc/pull/643/commits/7c99db64a9bb1cfc9e6efc81fd06a29b5be31d47